dev/Kozani/KozaniManager/ConnectionManager.cpp

```diff
@@ -0,0 +1,534 @@
+            {
+                THROW_LAST_ERROR_MSG("RegisterWaitForSingleObject failed. Cannot track lifetime of the associated local proess pid=%u", associatedLocalProcessId);
+            }
+
+            requestInfo->associatedLocalProcessHandle = std::move(associatedLocalProcessHandle);
+        }
+
+        m_requestsPendingConnection.push_back(requestInfo);
+
+        // Now that we are successful, disable removeRequestOnFailure to keep the newly added requestInfo in m_activationRequests list.
+        removeRequestOnFailure.release();
+        return requestInfo->statusReporter;
+    }
+
+    void ConnectionManager::ProcessProtocolDataUnit(
+            _In_reads_(size) BYTE* data,
+            UINT32 size,
+            _In_ IWTSVirtualChannelManager* channelManager,
+            _In_ IWTSVirtualChannel* channel)
+    {
+        Dvc::ProtocolDataUnit pdu;
+        if (!pdu.ParseFromArray(data, size))
+        {
+            LOG_HR_MSG(E_FAIL, "Failed to parse PDU. size: %u. Ignore failure.", size);
+            return;
+        }
+
+        Dvc::ProtocolDataUnit::DataType type{ pdu.type() };
+        switch (type)
+        {
+            case Dvc::ProtocolDataUnit::ConnectionAck:
+                ProcessConnectionAck(pdu, channelManager, channel);
+                break;
+
+            case Dvc::ProtocolDataUnit::ActivateAppResult:
+                ProcessActivateAppResult(pdu);
+                break;
+
+            case Dvc::ProtocolDataUnit::AppTerminationNotice:
+                ProcessAppTerminationNotice(pdu);
+                break;
+
+            default:
+                // ToDo: Unknown PDU type. Return GenericResult message with failure code to server side.
+                break;
+        }
+    }
+
+    // Must be called with m_requestsLock.lock_exclusive() done from the caller.
+    HRESULT ConnectionManager::OnDvcServerConnected(
+        UINT64 activityId,
+        ActivationRequestInfo* requestInfo,
+        _In_ IWTSVirtualChannelManager* channelManager,
+        _In_ IWTSVirtualChannel* channel,
+        _Out_ std::wstring& errorMessage) noexcept try
+    {
+        requestInfo->activityId = activityId;
+        requestInfo->dvcChannelManager = channelManager;
+        requestInfo->dvcChannel = channel;
+        requestInfo->status = RequestStatus::Connected;
+        requestInfo->statusReporter->SetStatus(RequestStatus::Connected);
+
+        auto result{ m_activityMap.insert({ activityId , requestInfo }) };
+        // activityId only increments. It will take hundreds of billion years for it to wrap around if it increments once every 1 second.
+        // So if there is a duplicate, there must be logical error in this code and we want to find out.
+        FAIL_FAST_IF_MSG(!result.second, "Activity ID should be unique for each request. Already existing ID: %I64u", activityId);
```

Comment:
> The caller of the function makes sure activityId >= 1.

A `RETURN_HR_IF(E_INVALIDARG, activityId == ActivationRequestInfo::ActivityId_Unknown);` is still useful in case future devs forgot about "the caller needs to make sure activity ID >= 1" and somehow passed 0 down here.

Reply:
E_UNEXPECTED (is there an E_INTERNAL?) is better as it's not merely an invalid parameter but "Hey, devs, you broke a 'this can never happen...and you can ensure it'" rule